### PR TITLE
rails 4 courtesy of Wardpenney 

### PR DESCRIPTION
**Pull request by phamdt**
...plus some resolved merges. tests pass.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -2,37 +2,30 @@ source 'https://rubygems.org'
 
 ruby '2.3.0'
 
-gem 'rails', '3.2.22.2'
-
+gem 'activeadmin', '1.0.0.pre2'
```
Comment (phamdt) [inline Gemfile]: upgraded

**Comment by phamdt** [comment]
oh, i found some text issues i need to work on

File: README.md [hunk 2]
```diff
@@ -15,13 +15,73 @@ Prelaunchr has a main mechanic from which everything else is derived: Every `Use
...
+
+`AdminUser.create!(:email => 'admin@example.com', :password => 'password', :password_confirmation => 'passwordconfirmaiton')`
+
+You can run this locally in a Rails console for development testing.
+
+If you are deployed to Heroku, you would run it there.
+=======
+View your website at the port default `http://localhost:7000/`.
+View sent mails at `http://localhost:1080/`.
+>>>>>>> 0b07c7c... Add information on viewing emails in development
```
Comment (irubnich) [inline README.md]: merge conflict

File: Procfile [hunk 3]
```diff
@@ -1,2 +1,3 @@
 web: bundle exec unicorn -p $PORT -c ./config/unicorn.rb
-worker:  bundle exec rake jobs:work
\ No newline at end of file
+worker:  bundle exec rake jobs:work
+mailcatcher: mailcatcher -f
```
Comment (irubnich) [inline Procfile]: Wouldn't this attempt to run `mailcatcher` in Heroku? I'd only put this locally in a `Procfile.dev` that can be run with `foreman -f Procfile.dev`

**Comment by irubnich** [comment]
Running migrations throws this deprecation warning:

```
W, [2016-03-04T21:53:02.127564 #8779]  WARN -- : DEPRECATION WARNING: `#timestamps` was called without specifying an option for `null`. In Rails 5, this behavior will change to `null: false`. You should manually specify `null: true` to prevent the behavior of your existing migrations from changing. (called from block in change at /Users/ilya/dev/prelaunchr/db/migrate/20130312045541_create_ip_addresses.rb:7)
```

Seems like an easy fix

**Comment by irubnich** [comment]
Setting the `config.ended` option to true in `application.rb` results in a redirect loop. There should definitely be a test for that as well

File: Gemfile (hunk 1 above)
Comment (irubnich) [inline Gemfile]: Awesome, and thanks for bumping everything else!

File: README.md [hunk 4]
```diff
@@ -15,13 +15,73 @@ Prelaunchr has a main mechanic from which everything else is derived: Every `Use
...
+bundle exec rake db:migrate
+```
+
+Start local server and mail worker:
+
+```no-highlight
+foreman start
+```
+
+<<<<<<< HEAD
```
Comment (irubnich) [inline README.md]: merge conflict

File: README.md [hunk 5]
```diff
@@ -15,13 +15,73 @@ Prelaunchr has a main mechanic from which everything else is derived: Every `Use
...
+
+In Rails console, run this command. Be careful to not use the example admin user
+for security reasons. Password confirmation should match password.
+
+`AdminUser.create!(:email => 'admin@example.com', :password => 'password', :password_confirmation => 'passwordconfirmaiton')`
+
+You can run this locally in a Rails console for development testing.
+
+If you are deployed to Heroku, you would run it there.
+=======
```
Comment (irubnich) [inline README.md]: merge conflict

File: config/environments/development.rb [hunk 6]
```diff
@@ -30,21 +34,14 @@
...
 
   # Expands the lines which load the assets
   config.assets.debug = true
 
-
   # For mailer configs
-  config.action_mailer.perform_deliveries = false
+  config.action_mailer.perform_deliveries = true
   config.action_mailer.raise_delivery_errors = true
+  config.action_mailer.default_url_options = { :host => ENV['DEFAULT_MAILER_HOST'] }
```
Comment (irubnich) [inline config/environments/development.rb]: The README should say that this needs to be configured on production.

File: app/models/admin_user.rb [hunk 7]
```diff
@@ -6,6 +6,5 @@ class AdminUser < ActiveRecord::Base
          :recoverable, :rememberable, :trackable, :validatable
 
   # Setup accessible (or protected) attributes for your model
-  attr_accessible :email, :password, :password_confirmation, :remember_me
```
Comment (irubnich) [inline app/models/admin_user.rb]: Please implement these as strong parameters if they're gonna be removed.

File: config/application.rb [hunk 8]
```diff
@@ -1,13 +1,10 @@
 require File.expand_path('../boot', __FILE__)
 
 require 'rails/all'
+require 'activeadmin'
```
Comment (irubnich) [inline config/application.rb]: These 2 requires aren't needed, Bundler takes care of all that.

**Comment by irubnich** [comment]
There's a blank file at the root called `prelaunchr`, not sure what it's there for. Kill it with  🔥

File: app/models/admin_user.rb (hunk 7 above)
Comment (phamdt) [inline app/models/admin_user.rb]: this isn't being called via a controller though so it's not exposed to receive arbitrary params from the world

File: db/migrate/20130126215239_create_users.rb [hunk 9]
```diff
@@ -5,7 +5,7 @@ def change
       t.string :referral_code
       t.integer :referrer_id
 
-      t.timestamps
+      t.timestamps null: true
```
Comment (phamdt) [inline db/migrate/20130126215239_create_users.rb]: i didn't love changing old migrations, but if the point is to get rid of warnings only, i suppose this works?

**Comment by phamdt** [comment]
putting campaign ending in env works for me, but it'd be even nicer if within the admin that could be toggled. that should be a separate PR though.

**Comment by irubnich** [comment]
Oh yea for sure, good call.

On Sun, Mar 06, 2016 at 1:16 PM phamdt

<
mailto:phamdt <notifications@github.com>
> wrote:

putting campaign ending in env works for me, but it'd be even nicer if within the admin that could be toggled. that should be a separate PR though.

**Comment by phamdt** [comment]
having issues with testing that redirect, so general suggestions are welcomed.

File: app/models/admin_user.rb (hunk 7 above)
Comment (irubnich) [inline app/models/admin_user.rb]: Oh true, that's great then. There wasn't any point to `attr_accessible` to begin with then...

**Comment by irubnich** [comment]
Thanks for the redirect fix, as for the test it should really just be a controller test for "the campaign is over, make sure it redirects to this path". I think RSpec would fail if there's a redirect loop.

**Comment by phamdt** [comment]
@irubnich, k, how bout now?

**Comment by phamdt** [comment]
i should probably change the other spec names then

**Comment by phamdt** [comment]
what, in this case, is the benefit of the stub?

**Comment by irubnich** [comment]
To avoid the `after` block - it just looks ugly to me honestly. I don't think there's any other benefit cause they're functionally equivalent.

**Comment by phamdt** [comment]
oh, that makes sense.

**Comment by phamdt** [comment]
ah, that doesn't work. it's setting a value not returning one so using `allow(Rails.application.config).to receive(:ended) { false }` doesn't work.

**Comment by ckhatton** [comment]
I guess soon Rails 5 will need to be thought about

**Comment by irubnich** [comment]
In the `UsersController` spec, I managed to replace the `before(:all)` with

``` ruby
before do
  allow(Rails.application.config).to receive(:ended) { false }
end
```

and it worked. Same with the integration spec, except returning `true`.

**Comment by phamdt** [comment]
ah, i tried in a before(:each), will try in the :all

**Comment by phamdt** [comment]
ping

**Comment by irubnich** [comment]
Great, merging this!

**Pull request merged**